File: src/sensors/switch/switch.ts
import { Subscriber } from '../../subscriber';
import { BinarySensor } from '../binary-sensor/binary-sensor';
import { HaDiscoverableManager } from '../../settings';
import { SwitchInfo } from './switch-info';

/**
 * Implements an MQTT switch for Home Assistant discovery
 */
export class Switch
  extends Subscriber<SwitchInfo, 'ON' | 'OFF'>
  implements Pick<BinarySensor, Exclude<keyof BinarySensor, 'entity' | keyof Subscriber<never>>>
{
  constructor(entity: SwitchInfo, settings: HaDiscoverableManager) {
    super(entity, settings);
  }

  switchOff() {
    return BinarySensor.prototype.switchOff.call(this);
  }

  switchOn() {
    return BinarySensor.prototype.switchOn.call(this);
  }

  updateState(state: boolean) {
    return BinarySensor.prototype.updateState.call(this, state);
  }
}

File: py-source/sensors/switch.py
# Inherit the on and off methods from the BinarySensor class, changing only the
# documentation string
from ha_mqtt_discoverable import Subscriber
from ha_mqtt_discoverable.sensors import SwitchInfo, BinarySensor


class Switch(Subscriber[SwitchInfo], BinarySensor):
    """Implements an MQTT switch:
    https://www.home-assistant.io/integrations/switch.mqtt
    """

    def off(self):
        """
        Set switch to off
        """
        super().off()

    def on(self):
        """
        Set switch to on
        """
        super().on()
